### PR TITLE
Try to reduce memory leakage and use (BL-2786)

### PR DESCRIPTION
**Pull request by StephenMcConnel**
Added a few more settings, and more debug window menu items.
Remove event handlers/listeners better when done with an html document.
Dispose images and gecko html objects that are no longer needed.

[Review on Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/826)

### CODE REVIEW

File: src/BloomBrowserUI/bookEdit/js/bloomEditing.js [hunk 1]
```diff
@@ -725,6 +725,18 @@ var pageSelectionChanging = function () {
     fireCSharpEditEvent('finishSavingPage', '');
 };
 
+// This is invoked from C# when we are about to leave a page (often right after the previous
+// method for changing pages).  It is mainly to clean things up so that garbage collection
+// won't lose multiple megabytes of data that both the DOM (C++) and Javascript subsystems
+// think the other is still using.
+var cleanupPageUponDeparture = function () {
+    var marginBox = $('.marginBox');
```
Comment (hatton) [inline src/BloomBrowserUI/bookEdit/js/bloomEditing.js]: Curious why you're starting at marginBox? We might well have an image attached to the page to provide full-bleed background.

Comment (StephenMcConnel) [inline src/BloomBrowserUI/bookEdit/js/bloomEditing.js]: I don't know that there's any reason for starting there.  I'd be happy to start it at a higher element (even body, i suppose).

**Comment by hatton** [comment]
Review status: 0 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[src/BloomBrowserUI/bookEdit/js/bloomEditing.js, line 732 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pVF7LwY9e9AtfGSVU)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomBrowserUI/bookEdit/js/bloomEditing.js#L732)):</sup>
We have another "cleanup()" method that gets called, so this should either be combined with that or get a different name.  How about "disconnectForGarbageCollection()"?


---

<sup>**[src/BloomBrowserUI/bookEdit/js/bloomEditing.js, line 738 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pTD5U5Q9wjSSWjqvy)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomBrowserUI/bookEdit/js/bloomEditing.js#L738)):</sup>
We also have images as background-image, and the above won't find those.  What you'll need to do is add  a bloomImages.js equivalent to HtmlDom.SelectChildImgAndBackgroundImageElements(). 

The closest we have in javascript is bloomImages.GetRawImageUrl().


---

<sup>**[src/BloomExe/Browser.cs, line 138 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pVlXsTJ0Pry9C89M-)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomExe/Browser.cs#L138)):</sup>
I'm somewhat uncomfortable with adding things that "might help". Is that the category you'd put these in? If so, how about we just commit these commented out.


---

<sup>**[src/BloomExe/Browser.cs, line 456 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pW2BuiVHh6HaI-C4V)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomExe/Browser.cs#L456)):</sup>
Let's limit all 3 of these to #if DEBUG


---

<sup>**[src/BloomExe/Edit/EditingModel.cs, line 663 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pWQlbE68B6chX0OaR)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomExe/Edit/EditingModel.cs#L663)):</sup>
What is "standard" about these event listeners? Hmmm. I think it's that they are more permanent than some others?


---

<sup>**[src/BloomExe/Edit/EditingModel.cs, line 675 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pXQnXyvrdlqB988PZ)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomExe/Edit/EditingModel.cs#L675)):</sup>
How about a new method that this can call, "AddMessageEventListener(name, Action<>)" which stores all the names as a byproduct. Then your RemoveStandardEventListeners can just loop over that list, disconnecting each one. That will make it less likely that in the future someone will add one without removing it.


---

<sup>**[src/BloomExe/Edit/WebThumbNailList.cs, line 497 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pX8kNQl1AKlFsb40a)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomExe/Edit/WebThumbNailList.cs#L497)):</sup>
Can this get a comment explaining why you want these off for a while?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826)

**Comment by hatton** [comment]
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826)

**Comment by StephenMcConnel** [comment]
Review status: 2 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[src/BloomBrowserUI/bookEdit/js/bloomEditing.js, line 732 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pVF7LwY9e9AtfGSVU)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomBrowserUI/bookEdit/js/bloomEditing.js#L732)):</sup>
done


---

<sup>**[src/BloomBrowserUI/bookEdit/js/bloomEditing.js, line 733 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1kDgP0z4xOhWhUzY18-r1-733)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomBrowserUI/bookEdit/js/bloomEditing.js#L733)):</sup>
Done.  (changed to use '.bloom-page')


---

<sup>**[src/BloomBrowserUI/bookEdit/js/bloomEditing.js, line 738 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pTD5U5Q9wjSSWjqvy)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomBrowserUI/bookEdit/js/bloomEditing.js#L738)):</sup>
Done.  (Although I didn't write a method in bloomImages since I'm not sure how to write it in javascript.)


---

<sup>**[src/BloomExe/Browser.cs, line 138 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pVlXsTJ0Pry9C89M-)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomExe/Browser.cs#L138)):</sup>
Done.


---

<sup>**[src/BloomExe/Browser.cs, line 456 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pW2BuiVHh6HaI-C4V)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomExe/Browser.cs#L456)):</sup>
I think they are, aren't they?


---

<sup>**[src/BloomExe/Edit/EditingModel.cs, line 663 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pWQlbE68B6chX0OaR)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomExe/Edit/EditingModel.cs#L663)):</sup>
"Standard" in that they're the ones that were being set every time by EditingModel.


---

<sup>**[src/BloomExe/Edit/EditingModel.cs, line 675 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pXQnXyvrdlqB988PZ)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomExe/Edit/EditingModel.cs#L675)):</sup>
Done.


---

<sup>**[src/BloomExe/Edit/WebThumbNailList.cs, line 497 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pX8kNQl1AKlFsb40a)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomExe/Edit/WebThumbNailList.cs#L497)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826)

**Comment by hatton** [comment]
Reviewed 5 of 6 files at r2.
Review status: 7 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[src/BloomBrowserUI/bookEdit/js/bloomEditing.js, line 738 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pTD5U5Q9wjSSWjqvy)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomBrowserUI/bookEdit/js/bloomEditing.js#L738)):</sup>
Hmmm, so this will work so long as there is the class you expect, but won't otherwise. Try this:
   
    $("[style*=background-image]")


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826)

**Comment by StephenMcConnel** [comment]
Review status: 6 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[src/BloomBrowserUI/bookEdit/js/bloomEditing.js, line 738 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pTD5U5Q9wjSSWjqvy)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomBrowserUI/bookEdit/js/bloomEditing.js#L738)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826)

**Comment by JohnThomson** [comment]
Review status: 6 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[src/BloomBrowserUI/bookEdit/js/bloomEditing.js, line 738 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pTD5U5Q9wjSSWjqvy)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomBrowserUI/bookEdit/js/bloomEditing.js#L738)):</sup>
Why get fancy? We're disconnecting in order to throw away this page! How about find().off() and get rid of every event function on everything? It may still be useful to be explicit about the things we need to delete...


---

<sup>**[src/BloomExe/Browser.cs, line 140 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1uqL-KGvJoqTbPsVV1)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/16c5b78ae22535c0f623d26203ef1ffefcb39241/src/BloomExe/Browser.cs#L140)):</sup>
That link seems to explain the first few, but to save time a one-line comment summarizing what each setting is supposed to do would help, I think. Many of them I can't even find a useful explanation by googling. If you found a decent reference that explains what each actually does that would be very helpful.

Is there any feasible way we could have our testers beat on these to see which ones make a significant difference (a) to the memory leak problem (b) to performance in different environments? Or have you definitely established that each is helpful (or at least not harmful)? "look like they might" does not build confidence.


---

<sup>**[src/BloomExe/Browser.cs, line 145 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1urKp3_1yuVJqH2c4w)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/16c5b78ae22535c0f623d26203ef1ffefcb39241/src/BloomExe/Browser.cs#L145)):</sup>
From what I can discover this only causes it to discard images in TABS that are inactive. We aren't using tabs.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826)

**Comment by hatton** [comment]
Reviewed 1 of 6 files at r2.
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[src/BloomBrowserUI/bookEdit/js/bloomEditing.js, line 738 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pTD5U5Q9wjSSWjqvy)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomBrowserUI/bookEdit/js/bloomEditing.js#L738)):</sup>
Steve, what do you say to JohnT's idea there of running this on every element. Would be nice if we could test how long it took.


---

<sup>**[src/BloomExe/Browser.cs, line 456 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pW2BuiVHh6HaI-C4V)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomExe/Browser.cs#L456)):</sup>
I don't see that in the code here. And I do see other commands that should not be DEBUG only, so... I'm gonna stick my neck out and reply "no" :-)


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826)

**Comment by StephenMcConnel** [comment]
Review status: 5 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[src/BloomBrowserUI/bookEdit/js/bloomEditing.js, line 738 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pTD5U5Q9wjSSWjqvy)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomBrowserUI/bookEdit/js/bloomEditing.js#L738)):</sup>
Done.


---

<sup>**[src/BloomExe/Browser.cs, line 456 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pW2BuiVHh6HaI-C4V)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomExe/Browser.cs#L456)):</sup>
Setting the three debug-only menu items is all enclosed inside #if DEBUG ... #endif.  Now if you want the method bodies also enclosed in a #if DEBUG block, I've done that now as well.  (But that's not what I understood you to be saying in the original comment.)


---

<sup>**[src/BloomExe/Browser.cs, line 140 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1uqL-KGvJoqTbPsVV1)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/16c5b78ae22535c0f623d26203ef1ffefcb39241/src/BloomExe/Browser.cs#L140)):</sup>
Done.  I tested the remaining new settings.  They all showed lower memory usage during the edit cycle of switching pages compared to not having them.


---

<sup>**[src/BloomExe/Browser.cs, line 145 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1urKp3_1yuVJqH2c4w)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/16c5b78ae22535c0f623d26203ef1ffefcb39241/src/BloomExe/Browser.cs#L145)):</sup>
Done.  (line removed--I found a reference backing up your claim)


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826)

**Comment by StephenMcConnel** [comment]
Review status: 5 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[src/BloomBrowserUI/bookEdit/js/bloomEditing.js, line 738 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pTD5U5Q9wjSSWjqvy)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomBrowserUI/bookEdit/js/bloomEditing.js#L738)):</sup>
I didn't try timing it, but the code isn't  noticeably slower with this change in it.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826)

**Comment by hatton** [comment]
LGTM


---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[src/BloomExe/Browser.cs, line 456 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826#-K1pW2BuiVHh6HaI-C4V)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/cc7ae3bcb68ce38ce3f12c28352d2f16bea29a6f/src/BloomExe/Browser.cs#L456)):</sup>
Hmmm confusing. No, I was just looking for the 3 commands to not be added. Sorry 'bout the confusion.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/826)

**Pull request merged**